local_db/api.py

```diff
@@ -62,3 +72,13 @@ def get_requests_for_user(self, user: User):
             if user.output_checker or metadata.workspace in user.workspaces:
                 requests.append(self._request(metadata))
         return requests
+
+    def set_status(self, request: ReleaseRequest, status: Status, user: User):
+        with transaction.atomic():
+            # validate transition, permissions, and update object
+            super().check_status(request, status, user)
```

Comment:
I split set_status into check_status and then set_status, to be able to handle this. The comment in the ProviderAPI is out of date - I will update it.. However, I absolutely want to run check_status as part of set_status, because a) its very cheap and b) it catches when a future developer (likely me!) forgets to call check_status first.

I really added the check_status for the `release_files` call, for which we have no transaction, and thus *cannot* roll back in the face of `set_status()` raising. And if we set_status *before* calling the old_api, and the old_api call errors, then the ReleaseRequest object is left mutated as if we had succeeded, and we'd need to undo. I thought a check/commit/set pattern was cleaner, even if I did not document that properly. I didn't actually spot this issue until this morning, its subtle.

For the local_db `set_status(...)` method implementation, we could just call set_status *after* updating the db, as the transaction will DTRT if it errors. But since check_status was available, I used it, to help provide an example of the pattern.

I will update the doc string for set_status to something like:

```
As calling set_status will mutate the passed ReleaseRequest, in cases where we may want to call to external services (e.g. job-server) to mutate external state, and these calls might fail, we provide a look-before-you-leap API. That is, when changing status and related state, an implementer should call `check_status(...)` first to validate the desired state transition and permissions are valid, then mutate their own state, and then call `set_status(...)` if successful to mutate the passed ReleaseRequest object.
```

Reply:
Makes sense, thanks for the explanation